src/main/kotlin/org/dif/crypto/Parse.kt

```diff
@@ -0,0 +1,52 @@
+package org.dif.crypto
+
+import com.nimbusds.jose.JWEObjectJSON
+import com.nimbusds.jose.JWSObjectJSON
+import com.nimbusds.jose.util.JSONObjectUtils
+import org.dif.common.Typ
+import org.dif.exceptions.DIDCommException
+import org.dif.exceptions.ParseException
+import org.dif.message.Message
+
+fun parse(str: String): ParseResult = try {
+    val json = JSONObjectUtils.parse(str)
+
+    when {
+        json.containsKey("signatures") -> ParseResult(Typ.Signed, JWSObjectJSON.parse(json))
+        json.containsKey("recipients") -> ParseResult(Typ.Encrypted, JWEObjectJSON.parse(json))
+        else -> ParseResult(Typ.Plaintext, Message.parse(json))
+    }
+} catch (t: Throwable) { throw ParseException("An error occurred while parsing the message", t) }
+
+class ParseResult {
+    private var _jwm: Message? = null
+    private var _jws: JWSObjectJSON? = null
+    private var _jwe: JWEObjectJSON? = null
+
+    var typ: Typ
+        private set
+
+    val jwm: Message
+        get() = _jwm ?: throw DIDCommException("JWM should be specified")
```

Comment:
Did you consider alternatives:
1) Algebraic Data Types (for example https://medium.com/sharenowtech/kotlin-adt-74472319962a)
2) Just separate classes for `ParseResultJWM`, `ParseResultJWE`, `ParseResultPlain`, and something like 
```
when (parseResult) {
    is ParseResultJWM -> ...
    is ParseResultJWE -> ...
    is ParseResultPlain -> ...
}
```